opencog/rule-engine/forwardchainer/ForwardChainer.cc

```diff
@@ -144,19 +147,50 @@ void ForwardChainer::do_chain(ForwardChainerCallBack& fcb,
      init_sources = _as.get_outgoing(hsource);
     else
         init_sources.push_back(hsource);
+
     _fcmem.update_potential_sources(init_sources);
 
-    _fcmem.set_source(fcb.choose_next_source(_fcmem)); //set initial source
     auto max_iter = _configReader.get_maximum_iterations();
 
     while (_iteration < max_iter /*OR other termination criteria*/) {
+
         _log->info("Iteration %d", _iteration);
 
-        do_step(fcb);
+        if(rule_weight.empty())
+        {
+            //! Choose next source.
+            _log->info("[ForwardChainer] setting next source");
+
+            _fcmem.set_source(choose_next_source(_fcmem));
+
+            // Choose matching rules whose input matches with the source.
+            vector<Rule*> matched_rules = choose_rules(_fcmem);
```

Comment:
I like the idea of having two separate classes and it used to exist in my initial design( separating ForwardChaining mechanism from Policies like DefaultChaining and R2Lchaining...whatever that comes in the future.) but should we not complete this revision first? the design could be considered on next iteration. Rather we should focus on making sure things work according to the specs.